File: frontend1/lib/services/stock_service.dart
import 'package:frontend1/services/api_service.dart';
import 'package:frontend1/models/medicine.dart';

/// Service pour la gestion du stock de médicaments
class StockService {
  final ApiService _apiService = ApiService();

  /// Récupère la liste paginée des médicaments
  /// [page] : Numéro de page (commence à 1)
  /// [search] : Terme de recherche optionnel
  Future<StockResponse> getMedicines({
    int page = 1,
    int limit = 10,
    String? search,
  }) async {
    try {
      final queryParams = <String, dynamic>{'page': page, 'limit': limit};
      if (search != null && search.isNotEmpty) {
        queryParams['search'] = search;
      }

      final response = await _apiService.get(
        '/stock/medicines',
        queryParameters: queryParams,
      );

      return StockResponse.fromJson(response.data as Map<String, dynamic>);
    } catch (e) {
      print('[StockService] Erreur getMedicines: $e');
      rethrow;
    }
  }

  /// Crée un nouveau médicament
  /// [data] : Map contenant tous les champs du formulaire
  Future<Medicine> createMedicine(Map<String, dynamic> data) async {
    try {
      final response = await _apiService.post('/stock/medicines', data: data);

      return Medicine.fromJson(response.data as Map<String, dynamic>);
    } catch (e) {
      print('[StockService] Erreur createMedicine: $e');
      rethrow;
    }
  }

  /// Met à jour un médicament existant
  /// [id] : ID du médicament
  /// [data] : Map contenant les champs à modifier
  Future<Medicine> updateMedicine(int id, Map<String, dynamic> data) async {
    try {
      final response = await _apiService.put(
        '/stock/medicines/$id',
        data: data,
      );

      return Medicine.fromJson(response.data as Map<String, dynamic>);
    } catch (e) {
      print('[StockService] Erreur updateMedicine: $e');
      rethrow;
    }
  }

  /// Supprime un médicament
  /// [id] : ID du médicament à supprimer
  Future<void> deleteMedicine(int id) async {
    try {
      await _apiService.delete('/stock/medicines/$id');
    } catch (e) {
      print('[StockService] Erreur deleteMedicine: $e');
      rethrow;
    }
  }
}

File: frontend/src/services/stockService.js
import api from './api';

export const stockService = {
  /**
   * Get paginated list of medicines with filters
   */
  async getMedicines(params = {}) {
    const { page = 1, pageSize = 50, search, familyId, typeId, isLowStock, isExpired } = params;
    
    const queryParams = new URLSearchParams({
      page: page.toString(),
      page_size: pageSize.toString(),
    });

    if (search) queryParams.append('search', search);
    if (familyId) queryParams.append('family_id', familyId.toString());
    if (typeId) queryParams.append('type_id', typeId.toString());
    if (isLowStock !== undefined) queryParams.append('is_low_stock', isLowStock.toString());
    if (isExpired !== undefined) queryParams.append('is_expired', isExpired.toString());

    const response = await api.get(`/stock/medicines?${queryParams.toString()}`);
    return response.data;
  },

  /**
   * Get a single medicine by ID
   */
  async getMedicine(id) {
    const response = await api.get(`/stock/medicines/${id}`);
    return response.data;
  },

  /**
   * Create a new medicine
   */
  async createMedicine(data) {
    const response = await api.post('/stock/medicines', data);
    return response.data;
  },

  /**
   * Update an existing medicine
   */
  async updateMedicine(id, data) {
    const response = await api.put(`/stock/medicines/${id}`, data);
    return response.data;
  },

  /**
   * Delete a medicine
   */
  async deleteMedicine(id) {
    await api.delete(`/stock/medicines/${id}`);
  },

  /**
   * Get stock alerts (low stock + expired)
   */
  async getStockAlerts() {
    const response = await api.get('/stock/alerts');
    return response.data;
  },
};

export default stockService;
